speechbrain/lobes/models/discrete/SQ-Codec.py

```diff
@@ -0,0 +1,1497 @@
+"""This lobe enables the integration of speech codec model (SQ-Codec) with scalar quantization,.
+
+SQ-Codec effectively maps the complex speech signal into a finite and compact latent space, named scalar latent space.
+
+Repository: https://github.com/yangdongchao/SimpleSpeech
+Paper: https://arxiv.org/abs/2406.02328, https://arxiv.org/abs/2408.13893
+
+Authors
+ * Pooneh Mousavi 2024
+"""
+
+import logging
+import os
+import zipfile
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torchaudio
+from huggingface_hub import hf_hub_download
+from omegaconf import OmegaConf
+from torch.autograd.function import InplaceFunction
+from torch.nn.utils import remove_weight_norm, weight_norm
+
+
+def download_and_extract(repo_id, filename, save_path):
+    """
+    Downloads a ZIP file from the specified repository, extracts its contents,
+    and removes the downloaded ZIP file.
+
+    Parameters
+    ----------
+    repo_id : str
+        The repository ID from which to download the ZIP file.
+    filename : str
+        The name of the file to download from the repository.
+    save_path : str
+        The directory where the contents of the ZIP file will be saved.
+    """
+    # Ensure save_path directory exists
+    if not os.path.exists(save_path):
+        os.makedirs(save_path)
+
+    # Download the file with progress bar
+    zip_filename = hf_hub_download(
+        repo_id=repo_id, filename=filename, cache_dir=save_path
+    )
+
+    # Extract the file
+    with zipfile.ZipFile(zip_filename, "r") as zip_ref:
+        zip_ref.extractall(save_path)
+
+    # Remove the downloaded ZIP file
+    os.remove(zip_filename)
+    print(f"File downloaded, extracted to '{save_path}', and ZIP file removed.")
+
+
+def decimal_to_ternary_matrix(decimals, D):
+    """
+    Convert a tensor of decimal numbers to a D*T ternary matrix for each batch.
+
+    Parameters
+    ----------
+    decimals : torch.Tensor
+        A 2D tensor of decimal numbers with shape (B, T), where B is the batch size
+        and T is the number of elements in each batch.
+    D : int
+        Number of ternary digits to represent each number (depth).
+
+    Returns
+    -------
+    torch.Tensor
+        A 3D tensor of shape (B, D, T) where each slice along the first dimension
+        corresponds to a batch, and each column is represented as a ternary number.
+    """
+    B, T = decimals.shape
+    ternary_matrix = torch.zeros((B, D, T), dtype=torch.long)
+    for pos in range(D):
+        ternary_matrix[:, pos, :] = decimals % 3  # Modulo operation
+        decimals //= 3  # Floor division for next ternary digit
+
+    return ternary_matrix
+
+
+def ternary_matrix_to_decimal(matrix):
```

Comment:
You will want to develop unittests for this function and put them in the test suit.